File: valid-word-abbreviation/Solution408.java
// https://leetcode.com/problems/valid-word-abbreviation/

class Solution408 {

    // T: O(n), S: O(1)
    public boolean validWordAbbreviation(String word, String abbr) {
        int wordLen = word.length(), abbrLen = abbr.length();
        int wordIdx = 0, abbrIdx = 0;
        while (wordIdx < wordLen && abbrIdx < abbrLen) {
            if (Character.isDigit(abbr.charAt(abbrIdx))) {
                if (abbr.charAt(abbrIdx) == '0') {
                    return false;
                }
                int jump = 0;
                while (abbrIdx < abbrLen && Character.isDigit(abbr.charAt(abbrIdx))) {
                    int digit = abbr.charAt(abbrIdx) - '0';
                    jump = jump * 10 + digit;
                    abbrIdx++;
                }
                wordIdx += jump;
            } else {
                if (wordIdx >= wordLen || word.charAt(wordIdx) != abbr.charAt(abbrIdx)) {
                    return false;
                }
                wordIdx++;
                abbrIdx++;
            }
        }
        return wordIdx == wordLen && abbrIdx == abbrLen;
    }

    public static void main(String[] args) {
        var sol = new Solution408();
        var w1 = "internationalization";
        var a1 = "i12iz4n";
        System.out.println(sol.validWordAbbreviation(w1, a1)); // true
        var w2 = "apple";
        var a2 = "a2e";
        System.out.println(sol.validWordAbbreviation(w2, a2)); // false
        var w3 = "internationalization";
        var a3 = "i5a11o1";
        System.out.println(sol.validWordAbbreviation(w3, a3)); // true
    }

}

File: valid-word-abbreviation/solution408.py
# https://leetcode.com/problems/valid-word-abbreviation/

class Solution408:

    # T: O(n), S: O(1)
    def validWordAbbreviation(self, word: str, abbr: str) -> bool:
        word_len, abbr_len = len(word), len(abbr)
        word_idx, abbr_idx = 0, 0

        while word_idx < word_len and abbr_idx < abbr_len:
            if abbr[abbr_idx].isdigit():
                if abbr[abbr_idx] == '0':
                    return False
                num = 0
                while abbr_idx < abbr_len and abbr[abbr_idx].isdigit():
                    digit = int(abbr[abbr_idx])
                    num = num * 10 + digit
                    abbr_idx += 1
                word_idx += num
            else:
                if word_idx > word_len or word[word_idx] != abbr[abbr_idx]:
                    return False
                word_idx += 1
                abbr_idx += 1

        return word_idx == word_len and abbr_idx == abbr_len


def main():
    sol = Solution408()
    w1, a1 = "internationalization", "i12iz4n"
    print(sol.validWordAbbreviation(w1, a1))  # True
    w2, a2 = "apple", "a2e"
    print(sol.validWordAbbreviation(w2, a2))  # False
    w3, a3 = "internationalization", "i5a11o1"
    print(sol.validWordAbbreviation(w3, a3))  # True


if __name__ == "__main__":
    main()
